Implement JobLifeCycle Object, use IDispoble for Task, Job, Stream.

jinlmsft opened the pull request:
This is a big aggregated change on the Prajna Project. This pull request is not fully ready to check-in yet. But I will use this pull request to initiate a code-review request, and
disucss the major changes incorporated in this version.

The main functions implemented are with regard to distributed exception handling and distributed resource management. Our goal will be: 1) Job may throw exception at remote, and all exception
will be serialized and shown at local. 2) We should properly implement handling of IDisposable object throughout our project.

1) JobLifeCycle object.

Each Prajna action (read, fold, etc.) is now associated with a GUID, which is further associated with a JobLifeCycle object. The JobLifeCycle governs the life cycle of a Prajna job,
and is responsible to release any resource used by Job (at App, Daemon, or at Container).

You can register callback (event) on the JobLifeCycle as follows.

OnDispose/OnDisposeFS: function to be called when the job is to be disposed. (only once right before job is disposed).
OnException/OnExceptionFS: function to be called when an exception is encountered. (may be called multiple times if multiple exception is encountered)
OnCancellation/OnCancellationFS: function to be called when the job is cancelled. (only once per job)

JobLifeCycle implements IDisposable interface.

2) SingleJobActionApp/SingleJobActionContainer/SingleJobActionDaemon

JobLifeCycle should not usually be used. Instead, function that plan to do something on the job should attempt to grab a SingleJobActionApp/SingleJobActionContainer/SingleJobActionDaemon object.
SingleJobAction\* object also implements IDisposable interface. If a SingleJobAction\* object has been successfully grabbed, the subsequent operation can safely use the resource of the job
without fearing that the resource is disposed. The operating logic is that when a SingleJobAction\* object will try to increment the reference count of JobLifeCycle when it is created, and
check whether the job is still properly running. If not, try to grab SingleJobAction\* object will return a null pointer, and the subsequent operations should not be executed.
When it is disposed, it will attempt to decrement the reference count. When Job is cancelled and the JobLifeCycle object is disposed, it will also decrement the reference count, and mark
the Job as being disposed. Only when the reference count goes to negative, the process will kick in to dispose the job resources.

3) ThrowExceptionAtCallback/EncounterExceptionAtCallback
   /ThrowExceptionAtContainer/EncounterExceptionAtContainer:

These functions are being used to catch exception at remote, aggregate exception, and throw at the local. Ultimately, we should move to a stage that most exceptions at remote (including our
own programming error) should be manifested at local.

4) Job/Task: disposable.

Both Job and Task become disposable object. Thus, they can properly free their resource if they fail.

5) MemStream/StreamBase<byte>, etc..

I start to use the disposable feature of MemStream/StreamBase<byte> to govern stream memory allocation and resource release. My thought is as follows.
a) Application that allocates the stream should be responsible to free its resource. That is, we should be able to
    use ms = new MemStream() ... for most of our jobs.
b) If other function needs to retain a copy of the stream, it should use .Replicate() to get a duplicated copy of the stream.
c) We should try to get rid of .AddRef() and .DecRef() management of the memory stream.

jinlmsft committed:
Implement JobLifeCycle Object, use IDispoble for Task, Job, Stream.

This is a big aggregated change on the Prajna Project. This pull request is not fully ready to check-in yet. But I will use this pull request to initiate a code-review request, and
disucss the major changes incorporated in this version.

The main functions implemented are with regard to distributed exception handling and distributed resource management. Our goal will be: 1) Job may throw exception at remote, and all exception
will be serialized and shown at local. 2) We should properly implement handling of IDisposable object throughout our project.

1) JobLifeCycle object.

Each Prajna action (read, fold, etc.) is now associated with a GUID, which is further associated with a JobLifeCycle object. The JobLifeCycle governs the life cycle of a Prajna job,
and is responsible to release any resource used by Job (at App, Daemon, or at Container).

You can register callback (event) on the JobLifeCycle as follows.

OnDispose/OnDisposeFS: function to be called when the job is to be disposed. (only once right before job is disposed).
OnException/OnExceptionFS: function to be called when an exception is encountered. (may be called multiple times if multiple exception is encountered)
OnCancellation/OnCancellationFS: function to be called when the job is cancelled. (only once per job)

JobLifeCycle implements IDisposable interface.

2) SingleJobActionApp/SingleJobActionContainer/SingleJobActionDaemon

JobLifeCycle should not usually be used. Instead, function that plan to do something on the job should attempt to grab a SingleJobActionApp/SingleJobActionContainer/SingleJobActionDaemon object.
SingleJobAction* object also implements IDisposable interface. If a SingleJobAction* object has been successfully grabbed, the subsequent operation can safely use the resource of the job
without fearing that the resource is disposed. The operating logic is that when a SingleJobAction* object will try to increment the reference count of JobLifeCycle when it is created, and
check whether the job is still properly running. If not, try to grab SingleJobAction* object will return a null pointer, and the subsequent operations should not be executed.
When it is disposed, it will attempt to decrement the reference count. When Job is cancelled and the JobLifeCycle object is disposed, it will also decrement the reference count, and mark
the Job as being disposed. Only when the reference count goes to negative, the process will kick in to dispose the job resources.

3) ThrowExceptionAtCallback/EncounterExceptionAtCallback
   /ThrowExceptionAtContainer/EncounterExceptionAtContainer:

These functions are being used to catch exception at remote, aggregate exception, and throw at the local. Ultimately, we should move to a stage that most exceptions at remote (including our
own programming error) should be manifested at local.

4) Job/Task: disposable.

Both Job and Task become disposable object. Thus, they can properly free their resource if they fail.

5) MemStream/StreamBase<byte>, etc..

I start to use the disposable feature of MemStream/StreamBase<byte> to govern stream memory allocation and resource release. My thought is as follows.
a) Application that allocates the stream should be responsible to free its resource. That is, we should be able to
	use ms = new MemStream() ... for most of our jobs.
b) If other function needs to retain a copy of the stream, it should use .Replicate() to get a duplicated copy of the stream.
c) We should try to get rid of .AddRef() and .DecRef() management of the memory stream.

zhuweirong commented on src/tools/tools/LoggerCSharp.fs:
```diff
@@ -47,6 +47,18 @@ type Logger private() =
 
+    /// Log message generated by "messageFunc" using "JobId", 
+    /// "messageFunc" is evaluated only if "logLevel" is less than or equal to the log level for "logId"
+    static member inline LogF(jobID : Guid, logLevel : LogLevel,messageFunc : Func<string>) =
```
This is to be used by C#, probably no need to "inline" to be consistent with other APIs.

jinlmsft replied:
Agree. And Will remove inline.

zhuweirong commented on src/tools/tools/Logger.fs:
```diff
@@ -297,6 +311,16 @@ type internal DefaultLogger () =
             // DefaultLogger does not support "log id" yet
             EmitLogEntry(logLevel, message)
+        
+        member this.Log ((logId : string, jobID: Guid, logLevel : LogLevel, message : string)) =
```
The "Logger" class below should have a non-functional flavor "Log" function that takes JobID as parameter

jinlmsft replied:
Revised. I do have a concern as the ILoggerProvider has 4 Log flavors that needed to be extended when someone implements a customized Log function. Will that be fine?

zhuweirong replied:
I don't expecting frequent extension though.

zhuweirong commented on src/tools/tools/StringTools.fs:
```diff
@@ -60,6 +61,16 @@ type StringTComparer<'T when 'T:equality>=
+/// <summary>
+/// Construct a comparer that uses Object.ReferenceEquals to compare object. 
+/// </summary>
+type ReferenceComparer<'T>()=
```
why it's in StringTools.fs? seems not specific to string.  Maybe in utils.fs?  Also, should it be internal?

zhuweirong commented on src/tools/tools/extension.fs:
```diff
@@ -356,4 +356,19 @@ type StringExtension =
+    /// See: http://blogs.msdn.com/b/brada/archive/2005/03/27/402801.aspx
+    [<Extension>]
+    static member WriteException( x: Stream, ex: Exception ) = 
+        let fmt = System.Runtime.Serialization.Formatters.Binary.BinaryFormatter()
```
Should it use the formatter that Bruno wrote?

jinlmsft replied:
For exception serialization (catch remote exception thrown), let's use BinaryFormatter(). The reason is that if the formatter Bruno wrote causes any issue, its exception can be captured and sent back in the Exception framework.

zhuweirong commented on src/tools/tools/extension.fs:
```diff
@@ -356,4 +356,19 @@ type StringExtension =
+    /// See: http://blogs.msdn.com/b/brada/archive/2005/03/27/402801.aspx
+    [<Extension>]
+    static member ReadException( x: Stream ) = 
+        let fmt = System.Runtime.Serialization.Formatters.Binary.BinaryFormatter()
```
Use the new formatter Bruno provided?

jinlmsft replied:
Reason above to use BinaryFormatter.

zhuweirong commented on src/CoreLib/DSeq.fs:
```diff
@@ -89,23 +90,26 @@ and internal DStreamForwardDependencyType =
-    // Retrieve a DStream
-    static member ResolveDStream( hash ) = 
-        DStreamFactory.Resolve( hash ) 
+    static let collectionByJob = ConcurrentDictionary<Guid,(int64 ref)*ConcurrentDictionary<_,_>>() 
```
add some comments to explain about the value type  (int64 ref) \* ConcurrentDictionary<_,_>, what are they?

jinlmsft replied:
```
/// A collection of DStream that serves as the network interface for intercommunication between Jobs. 
/// The entry is indexed by jobID, while each entry holds a timestamp (in ticks of int64), of the recent input from that job. 
/// That way, we can detect inactive jobs and cancel them later. 
```

zhuweirong commented on src/CoreLib/DSeq.fs:
```diff
@@ -532,71 +537,88 @@ and [<AllowNullLiteral>]
-        Logger.LogF( LogLevel.WildVerbose, ( fun _ -> sprintf "ProcessJobCommand, Map %A to peer %d" queuePeer.RemoteEndPoint peeri ))
-        try
+    member internal x.ProcessJobCommand( jobAction:SingleJobActionContainer, queuePeer:NetworkCommandQueue, cmd:ControllerCommand, ms:StreamBase<byte>, jobID: Guid ) = 
+//        using ( SingleJobActionContainer.TryFind(jobID)) ( fun jobAction -> 
```
remove the dead code?

jinlmsft replied:
Removed.

zhuweirong commented:
The UTs are not added yet?

zhuweirong commented on src/tools/tools/LoggerFSharp.fs:
```diff
@@ -50,6 +50,18 @@ type Logger private() =
+    /// "messageFunc" is evaluated only if "logLevel" is less than or equal to the log level for "logId"
+    static member inline LogF(logId: string, jobID : Guid, logLevel : LogLevel,messageFunc : unit -> string) =
+         if Logger.LoggerProvider.IsEnabled(logId, logLevel) then
+            Logger.LoggerProvider.Log((null,jobID, logLevel, messageFunc()))
```
null --> logId

jinlmsft replied:
Revised.

zhuweirong commented on src/tools/tools/Logger.fs:
```diff
@@ -82,6 +83,9 @@ type ILoggerProvider =
     /// Takes three parameters: log id, log level, log Message
     abstract member Log : (string*LogLevel*string) -> unit
 
+    /// Takes four parameters: log id, log level, JobID, log Message
```
comment does not match code

zhuweirong commented on src/tools/tools/Logger.fs:
```diff
@@ -82,6 +83,9 @@ type ILoggerProvider =
 
+    /// Takes four parameters: log id, log level, JobID, log Message
+    abstract member Log : (string*Guid*LogLevel*string) -> unit
+
```
To be consistent, should have an overload that  does (Guid \* LogLevel \* String) -> unit

jinlmsft replied:
Done.

zhuweirong commented on src/tools/tools/LoggerCSharp.fs:
```diff
@@ -47,6 +47,18 @@ type Logger private() =
+    /// "messageFunc" is evaluated only if "logLevel" is less than or equal to the log level for "logId"
+    static member inline LogF(logId: string, jobID : Guid, logLevel : LogLevel,messageFunc : Func<string>) =
+         if Logger.LoggerProvider.IsEnabled(logId, logLevel) then
+            Logger.LoggerProvider.Log((null,jobID, logLevel, messageFunc.Invoke()))
```
null -> logId

jinlmsft replied:
Done.

zhuweirong commented on src/tools/tools/LoggerCSharp.fs:
```diff
@@ -47,6 +47,18 @@ type Logger private() =
+    /// "messageFunc" is evaluated only if "logLevel" is less than or equal to the log level for "logId"
+    static member inline LogF(jobID : Guid, logLevel : LogLevel,messageFunc : Func<string>) =
+        if logLevel <= Logger.DefaultLogLevel then
+            Logger.LoggerProvider.Log((null,jobID, logLevel, messageFunc.Invoke()))
```
Once we have an overload of  "(Guid \* LogLevel \* String) -> unit", then there is no need to pass "null" for the first parameter.

jinlmsft replied:
Done.

zhuweirong commented on src/CoreLib/DSeq.fs:
```diff
@@ -532,71 +537,88 @@ and [<AllowNullLiteral>]
             | ( ControllerVerb.SyncWrite, ControllerNoun.DStream ) ->
                 let meta = BlobMetadata.Unpack( ms )
-                Logger.LogF( LogLevel.WildVerbose, ( fun _ -> sprintf "Rcvd %A command %A from peer %d with %dB" (meta.ToString()) cmd peeri (ms.Length-ms.Position) ))
+                Logger.LogF( jobID, LogLevel.MildVerbose, ( fun _ -> sprintf "Rcvd %A command %A from peer %d with %dB" (meta.ToString()) cmd peeri (ms.Length-ms.Position) ))
```
"Rcvd %A" --> "Rcvd %s", since meta.ToString() returns a string already.

jinlmsft replied:
Done.

zhuweirong commented on src/CoreLib/DSeq.fs:
```diff
@@ -532,71 +537,88 @@ and [<AllowNullLiteral>]
                 if bRet && (DateTime.UtcNow - queuePeer.LastSendTicks).TotalMilliseconds > 5000. then 
 //                if bRet then 
-                    let msEcho = new MemStream( 128 )
+                    use msEcho = new MemStream( 128 )
```
How the magic size "128" was determined?

jinlmsft replied:
An estimation of the size of the message. With the change Sanjeev will introduce, I assume that all new MemStream of a size small enough will be equivalent.

zhuweirong commented on src/CoreLib/DSeq.fs:
```diff
@@ -532,71 +537,88 @@ and [<AllowNullLiteral>]
-        | e -> 
-            Logger.Log( LogLevel.Info, ( sprintf "Error in processing DStream.Callback(%s:%s) cmd %A, peer %d, with exception %A" x.Name x.VersionString cmd peeri e )    )
-        true
+//                with 
```
remove dead code.

zhuweirong commented on src/CoreLib/DSeq.fs:
```diff
@@ -532,71 +537,88 @@ and [<AllowNullLiteral>]
+//                    true
+//        )
+    /// Parse echoing from DStream from a container to other container
+    member internal x.DStreamCallback( cmd, peeri, ms, jobID, name, verNumber ) =
```
DStream type is "internal" now. There is no need to mark the member "internal".

jinlmsft replied:
Internal removed.

zhuweirong commented on src/CoreLib/DSeq.fs:
```diff
@@ -532,71 +537,88 @@ and [<AllowNullLiteral>]
+    member internal x.DStreamCallback( cmd, peeri, ms, jobID, name, verNumber ) =
+        using ( SingleJobActionContainer.TryFind(jobID)) ( fun jobAction -> 
+            if Utils.IsNull jobAction then 
+                Logger.LogF( jobID, LogLevel.Info, ( fun _ -> sprintf "[May be Ok, Job cancelled?] DStream.DStreamCallback, received command %A of payload %dB, but can't find job lifecycle object" 
```
"Info" may get us a lot of log entries here.  In case of a DSet with many elements, and user only does a take 10.  It may take a while for all peers realize the cancellation. At the same time, we may get a lot of such log entries?

jinlmsft replied:
Good observation. I will change to MildVerbose.

zhuweirong commented on src/CoreLib/DSeq.fs:
```diff
@@ -673,23 +695,23 @@ and [<AllowNullLiteral>]
                 let msg = sprintf "DStream.SyncClose, DStream %s:%s, attempt to send SyncClose, DStream to peer %d, but the peer queue has already been shutdown" 
                             x.Name x.VersionString peeri 
-                Logger.Log( LogLevel.Warning, msg )
+                Logger.LogF( jbInfo.JobID, LogLevel.Warning, fun _ -> msg )
```
Not much benefit using LogF here.

Need to add a Log overload that takes JobID.

jinlmsft replied:
Change back to Log.

zhuweirong commented on src/CoreLib/DSeq.fs:
```diff
@@ -809,20 +834,20 @@ and [<AllowNullLiteral>]
                         Logger.Log( LogLevel.Error, msg )
                         let jbInfo = x.DefaultJobInfo
-                        let msError = new MemStream( 1024 )
+                        use msError = new MemStream( 1024 )
```
How the magic size "1024" was determined?

zhuweirong commented on src/CoreLib/DSetAction.fs:
```diff
@@ -71,26 +75,38 @@ type internal DSetAction() =
     /// Hold a Job object
-    member val Job  = null with get, set
+    member x.Job with get() = jobInternal
+                 and set(v: Job) = Volatile.Write( jobLifecycleRef, JobLifeCycleCollectionApp.BeginJob() )
```
why "Volatile.Write"?

jinlmsft replied:
Agree. Since each DSetAction only host a single job, I will change all Volatile.Write and .Read to normal reference access.

zhuweirong commented on src/CoreLib/DSetPeer.fs:
```diff
@@ -1013,12 +1025,16 @@ and [<AllowNullLiteral>]
-                        let msInfo = new MemStream( 10240 )
-                        msInfo.WriteString( msgInfo ) 
-                        x.HostQueue.ToSend( ControllerCommand( ControllerVerb.Warning, ControllerNoun.Message ) , msInfo )
+                        use msSend = new MemStream( 10240 )
```
why 10240?

jinlmsft replied:
use 1024.

zhuweirong commented on src/CoreLib/GV.fs:
```diff
@@ -360,16 +361,520 @@ and /// Information of Within Job cluster information.
+    let onException = ConcurrentQueue<Action<Exception>>()
+    let numExceptionCallback = ref 0
+    let exCollections = ConcurrentQueue<Exception>()
+    static member val MaxWaitToEndJobInMilliseconds = 30000 with get, set 
```
should it be made configurable in deploy.fs?

jinlmsft replied:
Agree, and add an entry in deploy.fs.

zhuweirong commented on src/CoreLib/GV.fs:
```diff
@@ -360,16 +361,520 @@ and /// Information of Within Job cluster information.
+    static member val MaxWaitToEndJobInMilliseconds = 30000 with get, set 
+    member val numJobActionsInProcess = ref 0 with get
+    /// Exception to throw here
+    member x.Exception with get() = if exCollections.IsEmpty then null else AggregateException( exCollections )
```
for internal types, we should consider using "option" instead of null.

jinlmsft replied:
Notice that JobLifeCycle and SingleJobAction are both disposable object. You can't use 
use a = new TypeA()

if TypeA is an option.

jinlmsft committed:
Ready to merge.

Add 3 UnitTest on Seq.take, Concurrent DSet read, remote exceution throwing.
Address Weirong's code review. Working with Sanjeev to merge new MemStream.

zhuweirong commented on tests/CoreLib/DSet.fs:
```diff
@@ -1536,3 +1536,70 @@ type DSetTests () =
+                    bExceptionCatched := false
+        | _ -> 
+            ()
+        Assert.AreEqual( !bExceptionCatched, true )
```
When this assertion fails, it's unclear it's because the expected exception is not caught, or it's because the msg does not match.

jinlmsft committed:
Make sure exception are thrown at main working thread.

Make sure exception are thrown at main working thread.
